File: src/main/java/app/tweditor/DBList.java
package app.tweditor;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class DBList extends DBElementValue
  implements Cloneable, Iterable<DBElement>
{
  private List<DBElement> elementList;
  private Map<String, DBElement> labelMap;
  private final AppEnvironment environment;

  public DBList(AppEnvironment environment, int capacity)
  {
    this.environment = environment;
    this.elementList = new ArrayList<>(capacity);
    this.labelMap = new HashMap<>(capacity);
  }

  public boolean addElement(DBElement element)
  {
    String label = element.getLabel();
    if (label.length() == 0) {
      this.elementList.add(element);
      return true;
    }

    if (this.labelMap.get(label) != null) {
      return false;
    }
    this.elementList.add(element);
    this.labelMap.put(label, element);
    return true;
  }

  public boolean insertElement(int index,DBElement element) {
    String label = element.getLabel();
    if (label.length() == 0) {
      this.elementList.add(index, element);
      return true;
    }

    if (this.labelMap.get(label) != null) {
      return false;
    }
    this.elementList.add(index, element);
    this.labelMap.put(label, element);
    return true;
  }

  public DBElement removeElement(int index)
  {
    DBElement element = this.elementList.get(index);
    this.elementList.remove(index);
    String label = element.getLabel();
    if (label.length() != 0) {
      this.labelMap.remove(label);
    }
    return element;
  }

  public boolean removeElement(String label)
  {
    if ((label == null) || (label.length() == 0)) {
      throw new IllegalArgumentException("No database element label supplied");
    }
    DBElement element = this.labelMap.get(label);
    if (element == null) {
      return false;
    }
    boolean removed = this.elementList.remove(element);
    if (removed) {
      this.labelMap.remove(label);
    }
    return removed;
  }

  public boolean removeElement(DBElement element)
  {
    boolean removed = this.elementList.remove(element);
    if (removed) {
      String label = element.getLabel();
      if (label.length() != 0) {
        this.labelMap.remove(label);
      }
    }
    return removed;
  }

  public DBElement getElement(String label)
  {
    if ((label == null) || (label.length() == 0)) {
      throw new IllegalArgumentException("No database element label supplied");
    }
    return this.labelMap.get(label);
  }

  public void setElement(String label, DBElement element)
  {
    if ((label == null) || (label.length() == 0)) {
      throw new IllegalArgumentException("No database element label supplied");
    }
    DBElement oldElement = this.labelMap.get(label);
    if (oldElement != null) {
      int index = this.elementList.indexOf(oldElement);
      this.elementList.set(index, element);
    } else {
      this.elementList.add(element);
    }

    this.labelMap.put(label, element);
  }

  public int getElementCount()
  {
    return this.elementList.size();
  }

  public Iterator<DBElement> iterator()
  {
    return this.elementList.iterator();
  }

  public DBElement getElement(int index)
  {
    return this.elementList.get(index);
  }

  public void setElement(int index, DBElement element)
  {
    DBElement oldElement = this.elementList.get(index);
    String oldLabel = oldElement.getLabel();
    String label = element.getLabel();
    if (!label.equals(oldLabel)) {
      throw new IllegalArgumentException("New label is not the same as old label");
    }
    this.elementList.set(index, element);
    this.labelMap.put(label, element);
  }

  public String getString(String label)
    throws DBException
  {
    DBElement element = getElement(label);
    String value;
    if (element != null) {
      int fieldType = element.getType();
      if (fieldType == 10) {
        value = (String)element.getValue();
      }
      else
      {
        if (fieldType == 11) {
          value = (String)element.getValue();
        }
        else
        {
          if (fieldType == 12) {
            LocalizedString string = (LocalizedString)element.getValue();
            if (string.getSubstringCount() > 0) {
              LocalizedSubstring substring = string.getSubstring(this.environment.getLanguageID(), 0);
              if (substring != null)
                value = substring.getString();
              else
                value = string.getSubstring(0).getString();
            } else {
              int refid = string.getStringReference();
              if (refid >= 0)
                value = this.environment.getString(refid);
              else
                value = "";
            }
          } else {
            throw new DBException("Field " + label + " is not a string");
          }
        }
      } } else { value = ""; }


    return value;
  }

  public void setString(String label, String value)
    throws DBException
  {
    DBElement element = getElement(label);
    if (element != null) {
      int fieldType = element.getType();
      if (fieldType == 10) {
        element.setValue(value);
      } else if (fieldType == 11) {
        element.setValue(value);
      } else if (fieldType == 12) {
        LocalizedString string = (LocalizedString)element.getValue();
        LocalizedSubstring substring = new LocalizedSubstring(value, this.environment.getLanguageID(), 0);
        string.addSubstring(substring);
      } else {
        throw new DBException("Field " + label + " is not a string");
      }
    } else {
      addElement(new DBElement(10, 0, label, value));
    }
  }

  public int getInteger(String label)
    throws DBException
  {
    DBElement element = getElement(label);
    int value;
    if (element != null) {
      int fieldType = element.getType();
      if ((fieldType == 0) || (fieldType == 2) || (fieldType == 3) || (fieldType == 5))
      {
        value = ((Integer)element.getValue()).intValue();
      }
      else
      {
        if ((fieldType == 6) || (fieldType == 7) || (fieldType == 4)) {
          value = ((Long)element.getValue()).intValue();
        }
        else
        {
          if (fieldType == 1) {
            value = ((Character)element.getValue()).charValue();
          }
          else
          {
            if (fieldType == 8) {
              value = ((Float)element.getValue()).intValue();
            }
            else
            {
              if (fieldType == 9)
                value = ((Double)element.getValue()).intValue();
              else
                throw new DBException("Field " + label + " is not numeric");  } 
          }
        }
      } } else { value = 0; }


    return value;
  }

  public void setInteger(String label, int value)
    throws DBException
  {
    setInteger(label, value, 5);
  }

  public void setInteger(String label, int value, int type)
    throws DBException
  {
    DBElement element = getElement(label);
    if (element != null) {
      int fieldType = element.getType();
      if (fieldType == 0) {
        element.setValue(value & 0xFF);
      } else if (fieldType == 2) {
        element.setValue(value & 0xFFFF);
      } else if (fieldType == 3) {
        int shortValue = value & 0xFFFF;
        if (shortValue > 32767)
          shortValue |= -65536;
        element.setValue(shortValue);
      } else if (fieldType == 5) {
        element.setValue(value);
      } else if (fieldType == 4) {
        element.setValue(Long.valueOf(value & 0xFFFFFFFF));
      } else if ((fieldType == 6) || (fieldType == 7)) {
        element.setValue(Long.valueOf(value));
      } else if (fieldType == 1) {
        element.setValue(Character.valueOf((char)value));
      } else if (fieldType == 8) {
        element.setValue(Float.valueOf(value));
      } else if (fieldType == 9) {
        element.setValue(Double.valueOf(value));
      } else {
        throw new DBException("Field " + label + " is not numeric");
      }
    } else {
      addElement(new DBElement(type, 0, label, value));
    }
  }

  public float getFloat(String label)
    throws DBException
  {
    DBElement element = getElement(label);
    float value;
    if (element != null) {
      int fieldType = element.getType();
      if (fieldType == 8)
        value = ((Float)element.getValue()).floatValue();
      else
        throw new DBException("Field " + label + " is not floating-point");
    }
    else {
      value = 0.0F;
    }

    return value;
  }

  public void setFloat(String label, float value)
    throws DBException
  {
    DBElement element = getElement(label);
    if (element != null) {
      int fieldType = element.getType();
      if (fieldType == 8)
        element.setValue(Float.valueOf(value));
      else
        throw new DBException("Field " + label + " is not floating-point");
    }
    else {
      addElement(new DBElement(8, 0, label, Float.valueOf(value)));
    }
  }

  public DBList clone()
  {
    DBList clonedList = (DBList)super.clone();

    int count = this.elementList.size();
    clonedList.elementList = new ArrayList<>(count);
    clonedList.labelMap = new HashMap<>(count);
    for (DBElement element : this.elementList) {
      clonedList.addElement(element.clone());
    }
    return clonedList;
  }
}

File: src/main/kotlin/app/tweditor/DBList.kt
package app.tweditor

class DBList(val environment: AppEnvironment, capacity: Int) : DBElementValue(), Iterable<DBElement> {
    var elementList: MutableList<DBElement> = ArrayList(capacity)
    var labelMap: MutableMap<String, DBElement> = HashMap(capacity)

    fun addElement(element: DBElement): Boolean {
        val label = element.getLabel()
        if (label.isEmpty()) {
            elementList.add(element)
            return true
        }

        if (labelMap[label] != null) {
            return false
        }
        elementList.add(element)
        labelMap[label] = element
        return true
    }

    fun insertElement(index: Int, element: DBElement): Boolean {
        val label = element.getLabel()
        if (label.isEmpty()) {
            elementList.add(index, element)
            return true
        }

        if (labelMap[label] != null) {
            return false
        }
        elementList.add(index, element)
        labelMap[label] = element
        return true
    }

    fun removeElement(index: Int): DBElement {
        val element = elementList.removeAt(index)
        val label = element.getLabel()
        if (label.isNotEmpty()) {
            labelMap.remove(label)
        }
        return element
    }

    fun removeElement(label: String?): Boolean {
        if (label == null || label.isEmpty()) {
            throw IllegalArgumentException("No database element label supplied")
        }
        val element = labelMap[label] ?: return false
        val removed = elementList.remove(element)
        if (removed) {
            labelMap.remove(label)
        }
        return removed
    }

    fun removeElement(element: DBElement): Boolean {
        val removed = elementList.remove(element)
        if (removed) {
            val label = element.getLabel()
            if (label.isNotEmpty()) {
                labelMap.remove(label)
            }
        }
        return removed
    }

    fun getElement(label: String?): DBElement? {
        if (label == null || label.isEmpty()) {
            throw IllegalArgumentException("No database element label supplied")
        }
        return labelMap[label]
    }

    fun setElement(label: String?, element: DBElement) {
        if (label == null || label.isEmpty()) {
            throw IllegalArgumentException("No database element label supplied")
        }
        val oldElement = labelMap[label]
        if (oldElement != null) {
            elementList[elementList.indexOf(oldElement)] = element
        } else {
            elementList.add(element)
        }

        labelMap[label] = element
    }

    fun setElement(index: Int, element: DBElement) {
        val oldElement = elementList[index]
        val oldLabel = oldElement.getLabel()
        val label = element.getLabel()
        if (label != oldLabel) {
            throw IllegalArgumentException("New label is not the same as old label")
        }
        elementList[index] = element
        labelMap[label] = element
    }

    fun getElement(index: Int): DBElement = elementList[index]

    fun getElementCount(): Int = elementList.size

    override fun iterator(): Iterator<DBElement> = elementList.iterator()

    @Throws(DBException::class)
    fun getString(label: String): String {
        val element = getElement(label)
        val value: String
        if (element != null) {
            val fieldType = element.getType()
            value = when (fieldType) {
                10, 11 -> element.getValue() as String
                12 -> {
                    val string = element.getValue() as LocalizedString
                    if (string.getSubstringCount() > 0) {
                        val substring = string.getSubstring(environment.languageID, 0)
                        if (substring != null) {
                            substring.string
                        } else {
                            string.getSubstring(0).string
                        }
                    } else {
                        val refid = string.stringReference
                        if (refid >= 0) {
                            environment.getString(refid)
                        } else {
                            ""
                        }
                    }
                }
                else -> throw DBException("Field " + label + " is not a string")
            }
        } else {
            value = ""
        }

        return value
    }

    @Throws(DBException::class)
    fun setString(label: String, value: String) {
        val element = getElement(label)
        if (element != null) {
            val fieldType = element.getType()
            when (fieldType) {
                10 -> element.setValue(value)
                11 -> element.setValue(value)
                12 -> {
                    val string = element.getValue() as LocalizedString
                    string.addSubstring(LocalizedSubstring(value, environment.languageID, 0))
                }
                else -> throw DBException("Field " + label + " is not a string")
            }
        } else {
            addElement(DBElement(10, 0, label, value))
        }
    }

    @Throws(DBException::class)
    fun getInteger(label: String): Int {
        val element = getElement(label)
        val value: Int
        if (element != null) {
            val fieldType = element.getType()
            value = when (fieldType) {
                0, 2, 3, 5 -> element.getValue() as Int
                6, 7, 4 -> (element.getValue() as Long).toInt()
                1 -> (element.getValue() as Char).code
                8 -> (element.getValue() as Float).toInt()
                9 -> (element.getValue() as Double).toInt()
                else -> throw DBException("Field " + label + " is not numeric")
            }
        } else {
            value = 0
        }

        return value
    }

    @Throws(DBException::class)
    fun setInteger(label: String, value: Int) {
        setInteger(label, value, 5)
    }

    @Throws(DBException::class)
    fun setInteger(label: String, value: Int, type: Int) {
        val element = getElement(label)
        if (element != null) {
            val fieldType = element.getType()
            when (fieldType) {
                0 -> element.setValue(value and 0xFF)
                2 -> element.setValue(value and 0xFFFF)
                3 -> {
                    var shortValue = value and 0xFFFF
                    if (shortValue > 32767) {
                        shortValue = shortValue or -65536
                    }
                    element.setValue(shortValue)
                }
                5 -> element.setValue(value)
                4 -> element.setValue(value.toLong())
                6, 7 -> element.setValue(value.toLong())
                1 -> element.setValue(value.toChar())
                8 -> element.setValue(value.toFloat())
                9 -> element.setValue(value.toDouble())
                else -> throw DBException("Field " + label + " is not numeric")
            }
        } else {
            addElement(DBElement(type, 0, label, value))
        }
    }

    @Throws(DBException::class)
    fun getFloat(label: String): Float {
        val element = getElement(label)
        val value: Float
        if (element != null) {
            val fieldType = element.getType()
            if (fieldType == 8) {
                value = element.getValue() as Float
            } else {
                throw DBException("Field " + label + " is not floating-point")
            }
        } else {
            value = 0.0f
        }

        return value
    }

    @Throws(DBException::class)
    fun setFloat(label: String, value: Float) {
        val element = getElement(label)
        if (element != null) {
            val fieldType = element.getType()
            if (fieldType == 8) {
                element.setValue(value)
            } else {
                throw DBException("Field " + label + " is not floating-point")
            }
        } else {
            addElement(DBElement(8, 0, label, value))
        }
    }

    public override fun clone(): DBList {
        val clonedList = super.clone() as DBList

        clonedList.elementList = ArrayList(elementList.size)
        clonedList.labelMap = HashMap(elementList.size)
        for (element in elementList) {
            clonedList.addElement(element.clone())
        }
        return clonedList
    }
}
